plugins/inputs/kinesis_consumer/kinesis_consumer.go

```diff
@@ -238,8 +260,47 @@ func (k *KinesisConsumer) Start(ac telegraf.Accumulator) error {
 	return nil
 }
 
+func gzipDecompress(data []byte) ([]byte, error) {
+	zipData, err := gzip.NewReader(bytes.NewReader(data))
+	if err != nil {
+		return nil, err
+	}
+	defer zipData.Close()
+	return ioutil.ReadAll(zipData)
+}
+
+func zlibDecompress(data []byte) ([]byte, error) {
+	zlibData, err := zlib.NewReader(bytes.NewReader(data))
+	if err != nil {
+		return nil, err
+	}
+	defer zlibData.Close()
+	return ioutil.ReadAll(zlibData)
+}
+
+func (k *KinesisConsumer) getDecompressionFunc(algo string) decompress {
+	switch algo {
+	case gzipDecompression:
+		return gzipDecompress
+	case zlibDecompression:
+		return zlibDecompress
+	default:
+		return func(data []byte) ([]byte, error) {
+			return data, nil
+		}
+	}
+}
+
+func (k *KinesisConsumer) getRecordData(r *consumer.Record) ([]byte, error) {
+	return k.getDecompressionFunc(k.DecompressWith)(r.Data)
+}
```

Comment:
How about determining the decompression function once in `Init()` and store it in `KinesisConsumer`, e.g. as `decompress`? Then call `k.decompress(r.Data)` in `onMessage()`...